test/helpers/position_assertions.cc

```diff
@@ -1187,44 +1205,58 @@ void SymbolSearchAssertion::checkAllForQuery(std::string query,
         auto &location = symbol->location;
         auto line = fmt::format("{}:{:{}}", location->uri, location->range->start->line, lineWidth);
         symbolsByLine[line].push_back(symbol);
+        assertionsByLine[line]; // create entry in assertionsByLine, if none exists yet.
     }
 
-    bool success = true;
-    // Show any expected symbols that weren't returned by LSP
+    int numUnfulfilled = 0;
+    int numUnexpected = 0;
+    // Look line-by-line for unmatched symbols and assertions
     for (auto entry : assertionsByLine) {
         auto lineInfo = entry.first;
-        auto expected = entry.second;
-        auto &actual = symbolsByLine[lineInfo];
-        if (expected.size() == actual.size()) {
-            continue;
+        auto assertionsOnLine = entry.second;
+        auto &symbolsOnLine = symbolsByLine[lineInfo];
+
+        // Report assertions with no matching symbol
+        for (auto &assertion : assertionsOnLine) {
+            bool foundMatchingSymbol = false;
+            for (auto &symbol : symbolsOnLine) {
+                if (assertion->matches(uriPrefix, symbol)) {
+                    foundMatchingSymbol = true;
+                    break;
+                }
+            }
+            if (foundMatchingSymbol) {
+                continue;
+            }
+            numUnfulfilled++;
+            addFailureAtLocationWithSource(assertion->getLocation(uriPrefix), sourceFileContents, uriPrefix,
+                                           fmt::format("Unfulfilled assertion: {}", assertion->toString()));
```

Comment:
This message isn't as clear as the old one. Can you explain why the assertion failed? (The old message explained exactly what happened.)

Reply:
```suggestion
                                           fmt::format("Did not find expected result for workspace/symbol query `{}`: {}", query, assertion->toString()));
```

Reply:
I can do this, sure – some of your suggested change in wording should go here, and some should go in the assertion's `toString` method.  I'll split it up accordingly.